doc/source/basics/runtime_extension.rst

```diff
@@ -0,0 +1,123 @@
+.. _basics-runtime-extension:
+
+:octicon:`plug` Runtime Extension
+=================================
+
+fairseq2 provides a flexible runtime extension system that allows you to extend its functionality without modifying the core codebase.
+It uses ``setuptools`` entry points to dynamically load and register extensions during initialization.
+
+Overview
+--------
+
+The extension system is built around a dependency injection container (learn more in :ref:`basics-design-philosophy`) that manages fairseq2's components.
+Through this system, you can:
+
+* Register new models and model cards
+* Add custom asset providers and validators
+* Extend the runtime context
+* And more...
+
+Basic Usage
+-----------
+
+Before using any fairseq2 APIs, you must initialize the framework with :meth:`fairseq2.init_fairseq2`:
+
+.. code-block:: python
+
+    from fairseq2 import init_fairseq2
+
+    init_fairseq2()
+
+Creating Extensions
+-------------------
+
+To create an extension, define a setup function:
+
+.. code-block:: python
+
+    from fairseq2.runtime.dependency import DependencyContainer
+
+    def setup_my_extension(container: DependencyContainer) -> None:
+        # Register your custom components here using `context`.
+        pass
+
+Registering Extensions
+----------------------
+
+Extensions are registered using setuptools entry points. You can configure them in either ``setup.py`` or ``pyproject.toml``:
+
+Using ``setup.py``:
+
+.. code-block:: python
+
+    setup(
+        name="my-fairseq2-extension",
+        entry_points={
+            "fairseq2.extension": [
+                "my_extension = my_package.module:setup_my_extension",
+            ],
+        },
+    )
+
+Using ``pyproject.toml``:
+
+.. code-block:: toml
+
+    [project.entry-points."fairseq2.extension"]
+    my_extension = "my_package.module:setup_my_extension"
+
+Extension Loading Process
+-------------------------
+
+When ``init_fairseq2()`` is called, the following steps occur:
+
+1. fairseq2 components are initialized
+2. All registered extensions are discovered via entry points
+3. Each extension's setup function is called
+
+Complete Example
+----------------
+
+Here's a complete example of implementing a fairseq2 extension:
+
+.. code-block:: python
+
+    from fairseq2.assets import AssetEnvironmentResolver
+    from fairseq2.composition import register_package_assets
+    from fairseq2.runtime.dependency import DependencyContainer
+
+
+    def setup_my_extension(container: DependencyContainer) -> None:
+
+        # Register package metadata provider for cards
+        register_package_assets(container, package="my_package.cards")
+
+        def resolve_cluster_name(resolver: DependencyResolver) -> str:
```

Comment:
I think as an example, it would be nicer to show a simpler concept like simply registering a model like:

```
register_model_family(container, name=xyz, factory=create_my_model, ...)
```

I think for most users cluster name resolution is a fairly edge/advanced case. Even in FAIR, this is handled by fairseq2-ext, so most users don't care about how clusters are determined.